### PR TITLE
Update HWIntrinsicInfo::lookupId to use a binary search

### PR DESCRIPTION
**Pull request by tannergooding**

### CODE REVIEW

**Comment by jkotas** [comment]
What is the experience if somebody adds a new intrinsic and gets the order wrong?

**Comment by tannergooding** [comment]
> What is the experience if somebody adds a new intrinsic and gets the order wrong?

It depends on how they get it wrong. Assertions exist for some, but not all scenarios at the moment. In the worst case they'll only see SPMI regressions and test failures stemming from intrinsics no longer being resolved.

More ideally we'd have a simple one time validation path. I don't think we have an existing spot to easily do such validation today, so I can probably just have do it in `lookupId`, tracking if its already been done or not (debug only, of course).

**Comment by jkotas** [comment]
> I can probably just have do it in lookupId, tracking if its already been done or not (debug only, of course).

Yes, I think it would be a good idea.

**Comment by tannergooding** [comment]
CC. @dotnet/jit-contrib, this should be ready for review.

File: src/coreclr/jit/hwintrinsic.cpp [hunk 1]
```diff
@@ -469,6 +469,215 @@ CorInfoType Compiler::getBaseJitTypeFromArgIfNeeded(NamedIntrinsic       intrins
...
+
+        prevInfo = info;
+    }
+}
+
+static void ValidateHWIntrinsicIsaRangeArray()
+{
+    size_t count = sizeof(hwintrinsicIsaRangeArray) / sizeof(hwintrinsicIsaRangeArray[0]);
+
+    for (size_t i = 0; i < count; i++)
```
Comment (jkotas) [inline src/coreclr/jit/hwintrinsic.cpp:673]:
```suggestion
    for (size_t i = 0; i < ARRAY_SIZE(hwintrinsicIsaRangeArray); i++)
```

File: src/coreclr/jit/hwintrinsic.cpp [hunk 2]
```diff
@@ -621,33 +855,50 @@ NamedIntrinsic HWIntrinsicInfo::lookupId(Compiler*         comp,
     }
 #endif
 
-    for (int i = 0; i < (NI_HW_INTRINSIC_END - NI_HW_INTRINSIC_START - 1); i++)
+    size_t isaIndex = static_cast<size_t>(isa) - 1;
+    assert(isaIndex < (sizeof(hwintrinsicIsaRangeArray) / sizeof(hwintrinsicIsaRangeArray[0])));
```
Comment (jkotas) [inline src/coreclr/jit/hwintrinsic.cpp:859]:
```suggestion
    assert(isaIndex < ARRAY_SIZE(hwintrinsicIsaRangeArray));
```

File: src/coreclr/jit/hwintrinsic.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/hwintrinsic.cpp:673]: 👍, wasn't aware we had this helper. I'm used to `_countof` or similar but that hadn't been available for our Unix build.

Comment (jakobbotsch) [reply src/coreclr/jit/hwintrinsic.cpp]: There is also `ArrLen` that we generally prefer to use in the JIT, but let's not rerun CI for that (looks like we have a small number of uses of `ARRAY_SIZE` anyway).

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/hwintrinsic.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/hwintrinsic.cpp]: Logged https://github.com/dotnet/runtime/issues/103908

**Pull request merged**